lib/extendedkey.ex

```diff
@@ -5,44 +5,60 @@ defmodule Bitcoinex.ExtendedKey do
   alias Bitcoinex.Secp256k1.{Params, Point, PrivateKey, Math}
   alias Bitcoinex.Base58
 
+  use Bitwise, only_operators: true
+
   defmodule DerivationPath do
     @moduledoc """
     Contains a list of integers representing a bip32 derivation path
     """
 
+    @min_non_hardened_child_num 0
+    # 2^31 - 1
+    @max_non_hardened_child_num 0x80000000 - 1
+    @min_hardened_child_num 0x80000000
+    # 2^32 - 1
+    @max_hardened_child_num 0x100000000 - 1
+
+    @type hardened_child_num ::
+            unquote(@min_hardened_child_num)..unquote(@max_hardened_child_num)
+    @type non_hardened_child_num ::
+            unquote(@min_non_hardened_child_num)..unquote(@max_non_hardened_child_num)
+
+    @type child_num :: hardened_child_num | non_hardened_child_num
+
     @type t :: %__MODULE__{
-            p: list
+            child_nums: list(child_num)
           }
 
     @enforce_keys [
-      :p
+      :child_nums
     ]
-    defstruct [:p]
+    defstruct [:child_nums]
 
-    @softcap Math.pow(2, 31)
-    @hardcap @softcap * @softcap
+    def max_non_hardened_child_num(), do: @max_non_hardened_child_num
+    def max_hardened_child_num(), do: @max_hardened_child_num
 
     @spec path_to_string(t()) :: String.t()
-    def path_to_string(%__MODULE__{p: path}), do: tpath_to_string(path, "")
+    def path_to_string(%__MODULE__{child_nums: path}), do: tpath_to_string(path, "")
 
     defp tpath_to_string([], path_acc), do: path_acc
     defp tpath_to_string(["m" | rest], path_acc), do: tpath_to_string(rest, "m/" <> path_acc)
 
     defp tpath_to_string([l | rest], path_acc) do
       cond do
-        l >= @hardcap ->
-          {:error, "index cannot be greater than 2^32-1"}
+        l >= @max_hardened_child_num ->
```

Comment:
@hardcap was 2^32. @max_hardened_child_num is 2^32-1. This needs to be fixed.

Reply:
I don't think so.  https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#extended-keys `The hardened child keys use indices 2^31 through 2^32-1.`

Reply:
Yes, this is inclusive I believe. @max_hardened_child_num = 2^32-1, so if we want to use this we have to replace `>=` with `>`. The ranges used in BIP32 are inclusive I believe.

Reply:
right. sorry I misread the original msg. fixed